File: leetcode/letterCombinations.js

//方法一
var f1=function(str1,str2){
    var arr = []
    for(var i=0;i<str1.length;i++){
        for(var j=0;j<str2.length;j++){
            arr.push(str1[i]+str2[j]); 
        }
    }
    return arr;
}
var letterCombinations = function(digits) {
    var dic=['abc','def','ghi','jkl','mno','pqrs','tuv','wxyz'];
    if(digits.length==0){
        return [];
    }if(digits.length==1){
        var arr=[];
        var str_1=dic[parseInt(digits)-2],
            len=str_1.length;
        for(var i=0; i<len; i++){
            arr.push(str_1[i]);
        }
        return arr
    }else{
        return f1(dic[parseInt(digits[0])-2],letterCombinations(digits.slice(1)))
    }
};

//方法二
let letterCombinations = function(digits) {
    if (digits === '') return [];
    let dict = {
        '2': 'abc',
        '3': 'def',
        '4': 'ghi',
        '5': 'jkl',
        '6': 'mno',
        '7': 'pqrs',
        '8': 'tuv',
        '9': 'wxyz'
    };
    let res = [''];
    for (let digit of digits) {
        let len = res.length;
        while (len-- > 0) {
            let e = res.shift();
            for (let c of dict[digit]) {
                res.push(e + c);
            }
        }
    }
    return res;
};

File: leetcode/letterCombinations.py
# -*- coding: utf-8 -*-
#给定一个仅包含数字 2-9 的字符串，返回所有它能表示的字母组合。给出数字到字母的映射如下（与电话按键相同）。注意 1 不对应任何字母。
"""
Created on Tue Dec 18 13:05:24 2018

@author: make
"""
#方法一：
def f1(strs1,strs2):
    arr=[]
    for i in strs1:
        for j in strs2:
            arr.append(i+j)
    return arr

def letterCombinations(digits):
    dic={'2':'abc','3':'def','4':'ghi','5':'jkl','6':'mno','7':'pqrs','8':'tuv','9':'wxyz'}
    if len(digits)==0:
        return []
    elif len(digits)==1:
        return list(dic[digits])
    else:
        return f1(dic[digits[0]],letterCombinations(digits[1:]))



digits='23'
print(letterCombinations(digits))



#方法二

def letterCombinations(digits):
    dic={'2':'abc','3':'def','4':'ghi','5':'jkl','6':'mno','7':'pqrs','8':'tuv','9':'wxyz'}
    if len(digits)==0:
        return []
    acc=[e for e in dic[digits[0]]]
    for i in digits[1:]:
        acc=[w+k for w in acc for k in dic[i]]
    return acc
digits='23'
print(letterCombinations(digits))
